### ISSUE TITLE
Pipelining with interfaces with private data?

### ISSUE THREAD

**Issue by m1cr0man** [capnproto/capnproto-rust#298]
Hello. This probably falls more into general capnproto usage but since I'm implementing in Rust I figured this was the best place to ask 🙂 

I'm trying to figure out how to implement the Directory.link method shown in the [capnproto language examples](https://capnproto.org/language.html#interfaces). This example of pipelining is what drew me in to try capnproto in the first place. However, I can't figure out how to pull it off.

On the server side, I have a `NodeImpl` struct with a `path` attribute for use on server side operations.

```rust
use std::path::Path;

use capnp::capability::Promise;

use crate::schema::nodes::node_capnp::node;

pub struct NodeImpl<'a> {
    path: &'a Path,
}

impl<'a> NodeImpl<'a> {
    fn new(path: &'a Path) -> NodeImpl<'a> {
        NodeImpl { path }
    }
}
```

From this it's easy to implement any functions that operate on this node alone. The problem arises with the `link` function (for berevity, I have put the link method definition on the Node in my schema, so no Directory interface):

```rust
impl<'a> node::Server for NodeImpl<'a> {
    fn link(
        &mut self,
        params: node::LinkParams,
        _results: node::LinkResults,
    ) -> Promise<(), capnp::Error> {
        Promise::from_future(async move {
            let params = params.get()?;
            let name = params.get_name()?;
            let link_name = self.path.parent().unwrap().join(name);
            let target_node = params.get_node()?;

            // This isn't possible because target_node is a client
            let target_path = target_node.path;

            tokio::fs::symlink(target_path, link_name).await?;
            Ok(())
        })
    }
}
```

Now my function has received a client for a Node in its parameters. I can't figure out how I would dereference this back to something useful on the server side. I thought #172 would be fruitful but it seems that issue went in a different direction?

Instead, I looked into the [calculator example](https://github.com/capnproto/capnproto-rust/blob/master/capnp-rpc/examples/calculator/server.rs#L63) and gathered that I would need to implement a `path` function on the Node interface and call it in my link method:

```rust
        Promise::from_future(async move {
            let params = params.get()?;
            let name = params.get_name()?;
            let link_name = self.path.parent().unwrap().join(name);
            let target_node = params.get_node()?;

            let target_path = target_node
                .path_request()
                .send()
                .promise
                .await?
                .get()?
                .get_path()?;

            tokio::fs::symlink(target_path, link_name).await?;
            Ok(())
        })
```

Is this the only solution to the problem? Are there any ways for me to implement private functions or attributes on the Client? This solution feels really cumbersome and I don't like the fact that I need to expose the underlying implementation (aka `path`) to the client. If I had to chose, I'd prefer to not expose the raw path than deal with a long resolve method.

TIA

**Comment by dwrensha** [comment]
Yeah, #172 looks like the relevant issue here. capnproto-rust needs to provide support for something like capnproto-c++'s `CapabilityServerSet`.

**Comment by dwrensha** [comment]
#299 is a draft of what I had in mind. There are still some details to work out.

**Comment by m1cr0man** [comment]
Hey thanks for the info and doing up a draft! Was not expecting such a quick response. I'm not in any rush personally.

I still have more reading to do on CapabilityServerSet as I don't fully understand how it would be used. If I do gather enough understanding to help implement it I certainly will do what I can.

**Comment by dwrensha** [comment]
Closed by #299.
